[InstSimplify] Generalize simplification of icmps with monotonic operands

nikic opened the pull request:
InstSimplify currently folds patterns like `(x | y) uge x`  and `(x & y) ule x` to true. However, it cannot handle combinations of such situations, such as `(x | y) uge (x & z)` etc.

To support this, recursively collect operands of monotonic instructions (that preserve either a greater-or-equal or less-or-equal relationship) and then check whether any of them match.

Fixes https://github.com/llvm/llvm-project/issues/69333.

nikic committed:
fold monotonic icmps

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -3103,6 +3103,54 @@ static Value *simplifyICmpWithConstant(CmpInst::Predicate Pred, Value *LHS,
 }
 
+/// Get values V_i such that V uge V_i (Greater) or V ule V_i (!Greater).
+static void getUnsignedMonotonicValues(SmallPtrSetImpl<Value *> &Res, Value *V,
```
Prefer ptr to output reference.

goldsteinn replied:
On that note, however, do you actually need to collect the values? Based on the usage of just simplifying booleans, can't you just pass a query (greater, greater_eq, less_eq, less) and return a bool if its provable?

nikic replied:
Values are collected into the set in recursive calls, so this can't be a return value.

goldsteinn replied:
err, yeah realized err in my prior comment (deleted shortly after for context).

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -3103,6 +3103,54 @@ static Value *simplifyICmpWithConstant(CmpInst::Predicate Pred, Value *LHS,
 
+/// Get values V_i such that V uge V_i (Greater) or V ule V_i (!Greater).
+static void getUnsignedMonotonicValues(SmallPtrSetImpl<Value *> &Res, Value *V,
+                                       bool Greater, unsigned Depth = 0) {
```
Would replace greater with an enum that also support `GreaterOrEq` and `LessOrEq`. Also think a bool for `Signed` makes sense.

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -3103,6 +3103,54 @@ static Value *simplifyICmpWithConstant(CmpInst::Predicate Pred, Value *LHS,
+  Value *X, *Y;
+  if (Greater) {
+    if (match(V, m_Or(m_Value(X), m_Value(Y))) ||
+        match(V, m_Intrinsic<Intrinsic::uadd_sat>(m_Value(X), m_Value(Y)))) {
```
You can imagine a lot more cases for this. Imo should just be a switch from the start.

goldsteinn replied:
Also these matches aren't `Greater`, they are `GreaterOrEq`.

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -3103,6 +3103,54 @@ static Value *simplifyICmpWithConstant(CmpInst::Predicate Pred, Value *LHS,
+               match(V, m_UDiv(m_Value(X), m_Value())) ||
+               match(V, m_LShr(m_Value(X), m_Value())) ||
+               match(V, m_Intrinsic<Intrinsic::usub_sat>(m_Value(X)))) {
+      getUnsignedMonotonicValues(Res, X, Greater, Depth);
```
Don't need complete coverage here, but can you add TODO for some additional freebies (min/max/add+nuw/sub+nuw/mul+nuw/shl+nuw/ashr+signbit/etc...)

nikic commented:
Some crashes on llvm-opt-benchmark.

nikic committed:
Fix assertion failure

Given that it's easier to handle than not handle these cases,
avoid the order-dependence of transforms.

nikic commented:
Fixed the crash.

dtcxzyw reviewed (approved):
LGTM.
